### PR TITLE
Add cell diversity feature

### PR DESCRIPTION
**Pull request by camisowers**
**If you haven't already, please read through our [contributing guidelines](https://ark-analysis.readthedocs.io/en/latest/_rtd/contributing.html) before opening your PR**

**What is the purpose of this PR?**

Incorporate cell diversity into the SpaceCat class. 

**How did you implement your changes**

Use [squidpy.gr.spatial_neighbors()](https://squidpy.readthedocs.io/en/stable/api/squidpy.gr.spatial_neighbors.html) to generate a sparse cell distance matrix and calculate the neighborhood matrices from there.
Once, we have the frequencies matrix, it's simple to determine the diversity (Shannon Index) of each cell neighborhood.
This gets done at only one feature level, the same as the functional marker features. 

**Remaining issues**
Originally, I had intended to generate the neighborhood matrix in the preprocessing step, but since specifications like `image_key`, `seg_label_key`, `cluster_key`, etc. are needed it makes more sense to just do the calculation in the feature class.

The distance/neighborhood matrix generation currently takes around 13 minutes to run on the TONIC dataset (678 images, ~2 million cells). @jranek and I have discussed some potential speed ups using parallelization, which we can always go back and add in later.

### CODE REVIEW

**Review by ngreenwald — APPROVED** [review]
Looks good!

**Comment by camisowers** [comment]
Any progress on the neighbors matrix parallelization @jranek?

File: src/SpaceCat/features.py [hunk 1]
```diff
@@ -460,6 +461,120 @@ def generate_stats(self, table, params, df_name, var_name, filter_stats, dedupli
...
+            generates and saves feature dataframe, as well as the filtered dataframe
+        """
+        # create neighbor counts and frequencies matrix
+        self.create_neighborhood_matrix(diversity_feature_level, pixel_radius)
+
+        # calculate Shannon diversity per cell
+        cell_diversity_table = self.compute_neighborhood_diversity(diversity_feature_level, pixel_radius)
+
+        params = [[diversity_feature_level + '_freq', diversity_feature_level]]
+        self.generate_stats(cell_diversity_table, params, 'diversity_stats', 'cell_diversity',
```
Comment (jranek) [inline src/SpaceCat/features.py:571]: This is throwing some dtype categorical error for the Stanford cohort
```bash
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[63], line 8
      3 adata_processed.obs_names = [f'{i}' for i in range(0, len(adata_processed.obs_names))]
      4 features = SpaceCat(adata_processed, image_key='fov', seg_label_key='label', cell_area_key='area',
      5                     cluster_key=['cell_cluster', 'cell_cluster_broad'], 
      6                     compartment_key=None, compartment_area_key='compartment_area')
----> 8 adata_processed = features.run_spacecat(functional_feature_level='cell_cluster', diversity_feature_level='cell_cluster', pixel_radius = 50,
      9                                         per_cell_stats=per_cell_stats, per_img_stats=per_img_stats)
     12 # # Generate features and save anndata table
     13 # adata_processed = features.run_spacecat()
     14 # adata_processed.write_h5ad(os.path.join(output_dir, 'adata', 'adata_table_processed.h5ad'))

File ~/Documents/Angelo_lab/SpaceCat/features.py:811, in SpaceCat.run_spacecat(self, functional_feature_level, diversity_feature_level, pixel_radius, per_cell_stats, per_img_stats, filter_stats, deduplicate_stats, correlation_filtering_thresh)
    808 self.generate_abundance_features(stats_df, density_params, ratio_cluster_key=broadest_cluster_col)
    810 # generate diversity features
--> 811 self.generate_diversity_features(diversity_feature_level, pixel_radius, filter_stats)
    813 # add functional features to list
    814 per_cell_stats.append(['functional_marker', functional_feature_level, None])

File ~/Documents/Angelo_lab/SpaceCat/features.py:571, in SpaceCat.generate_diversity_features(self, diversity_feature_level, pixel_radius, filter_stats)
    568 cell_diversity_table = self.compute_neighborhood_diversity(diversity_feature_level, pixel_radius)
    570 params = [[diversity_feature_level + '_freq', diversity_feature_level]]
--> 571 self.generate_stats(cell_diversity_table, params, 'diversity_stats', 'cell_diversity',
    572                     filter_stats=filter_stats, deduplicate_stats=False)
    574 # format features
    575 df_name = 'diversity_stats_filtered' if filter_stats else 'diversity_stats'

File ~/Documents/Angelo_lab/SpaceCat/features.py:445, in SpaceCat.generate_stats(self, table, params, df_name, var_name, filter_stats, deduplicate_stats)
    442     cluster_names.remove(cluster_col_name)
    443     drop_cols.extend(cluster_names)
--> 445     stats_dfs.append(self.create_long_df(table=table,
    446                                          result_name=result_name,
    447                                          var_name=var_name,
    448                                          cluster_col_name=cluster_col_name,
    449                                          drop_cols=drop_cols,
    450                                          normalize=normalize,
    451                                          subset_col=subset_col))
    452 stats_df_comb = pd.concat(stats_dfs, axis=0)
    453 stats_df_comb['cell_type'] = stats_df_comb['cell_type'].astype(str)

File ~/Documents/Angelo_lab/SpaceCat/features.py:209, in SpaceCat.create_long_df(self, table, cluster_col_name, result_name, var_name, subset_col, cluster_stats, normalize, drop_cols)
    207 # first generate df without subsetting
    208 drop_cols_all = drop_cols + [subset_col] if subset_col is not None else drop_cols.copy()
--> 209 long_df_all = self.long_df_helper(table, cluster_col_name, drop_cols_all, var_name,
    210                                   cluster_stats, normalize)
    211 long_df_all['metric'] = result_name
    212 long_df_all['subset'] = 'all'

File ~/Documents/Angelo_lab/SpaceCat/features.py:176, in SpaceCat.long_df_helper(self, table, cluster_col_name, drop_cols, var_name, cluster_stats, normalize, subset_col)
    172         transformed = self.get_frequencies(transformed, groupby_cols)
    174 else:
    175     # get sum or average of values
--> 176     transformed = grouped_table.agg('mean') if normalize else grouped_table.agg('sum')
    177     transformed.reset_index(inplace=True)
    179     # reshape to long df

File ~/miniconda3/envs/venv_quiche/lib/python3.9/site-packages/pandas/core/groupby/generic.py:1432, in DataFrameGroupBy.aggregate(self, func, engine, engine_kwargs, *args, **kwargs)
   1429     kwargs["engine_kwargs"] = engine_kwargs
   1431 op = GroupByApply(self, func, args=args, kwargs=kwargs)
-> 1432 result = op.agg()
   1433 if not is_dict_like(func) and result is not None:
   1434     # GH #52849
   1435     if not self.as_index and is_list_like(func):

File ~/miniconda3/envs/venv_quiche/lib/python3.9/site-packages/pandas/core/apply.py:187, in Apply.agg(self)
    184 kwargs = self.kwargs
    186 if isinstance(func, str):
--> 187     return self.apply_str()
    189 if is_dict_like(func):
    190     return self.agg_dict_like()

File ~/miniconda3/envs/venv_quiche/lib/python3.9/site-packages/pandas/core/apply.py:603, in Apply.apply_str(self)
    601         else:
    602             self.kwargs["axis"] = self.axis
--> 603 return self._apply_str(obj, func, *self.args, **self.kwargs)

File ~/miniconda3/envs/venv_quiche/lib/python3.9/site-packages/pandas/core/apply.py:693, in Apply._apply_str(self, obj, func, *args, **kwargs)
    691 f = getattr(obj, func)
    692 if callable(f):
--> 693     return f(*args, **kwargs)
    695 # people may aggregate on a non-callable attribute
    696 # but don't let them think they can pass args to it
    697 assert len(args) == 0

File ~/miniconda3/envs/venv_quiche/lib/python3.9/site-packages/pandas/core/groupby/groupby.py:2452, in GroupBy.mean(self, numeric_only, engine, engine_kwargs)
   2445     return self._numba_agg_general(
   2446         grouped_mean,
   2447         executor.float_dtype_mapping,
   2448         engine_kwargs,
   2449         min_periods=0,
   2450     )
   2451 else:
-> 2452     result = self._cython_agg_general(
   2453         "mean",
   2454         alt=lambda x: Series(x, copy=False).mean(numeric_only=numeric_only),
   2455         numeric_only=numeric_only,
   2456     )
   2457     return result.__finalize__(self.obj, method="groupby")

File ~/miniconda3/envs/venv_quiche/lib/python3.9/site-packages/pandas/core/groupby/groupby.py:1998, in GroupBy._cython_agg_general(self, how, alt, numeric_only, min_count, **kwargs)
   1995     result = self._agg_py_fallback(how, values, ndim=data.ndim, alt=alt)
   1996     return result
-> 1998 new_mgr = data.grouped_reduce(array_func)
   1999 res = self._wrap_agged_manager(new_mgr)
   2000 if how in ["idxmin", "idxmax"]:

File ~/miniconda3/envs/venv_quiche/lib/python3.9/site-packages/pandas/core/internals/managers.py:1472, in BlockManager.grouped_reduce(self, func)
   1470             result_blocks = extend_blocks(applied, result_blocks)
   1471     else:
-> 1472         applied = blk.apply(func)
   1473         result_blocks = extend_blocks(applied, result_blocks)
   1475 if len(result_blocks) == 0:

File ~/miniconda3/envs/venv_quiche/lib/python3.9/site-packages/pandas/core/internals/blocks.py:393, in Block.apply(self, func, **kwargs)
    387 @final
    388 def apply(self, func, **kwargs) -> list[Block]:
    389     """
    390     apply the function to my values; return a block if we are not
    391     one
    392     """
--> 393     result = func(self.values, **kwargs)
    395     result = maybe_coerce_values(result)
    396     return self._split_op_result(result)

File ~/miniconda3/envs/venv_quiche/lib/python3.9/site-packages/pandas/core/groupby/groupby.py:1973, in GroupBy._cython_agg_general.<locals>.array_func(values)
   1971 def array_func(values: ArrayLike) -> ArrayLike:
   1972     try:
-> 1973         result = self._grouper._cython_operation(
   1974             "aggregate",
   1975             values,
   1976             how,
   1977             axis=data.ndim - 1,
   1978             min_count=min_count,
   1979             **kwargs,
   1980         )
   1981     except NotImplementedError:
   1982         # generally if we have numeric_only=False
   1983         # and non-applicable functions
   1984         # try to python agg
   1985         # TODO: shouldn't min_count matter?
   1986         # TODO: avoid special casing SparseArray here
   1987         if how in ["any", "all"] and isinstance(values, SparseArray):

File ~/miniconda3/envs/venv_quiche/lib/python3.9/site-packages/pandas/core/groupby/ops.py:831, in BaseGrouper._cython_operation(self, kind, values, how, axis, min_count, **kwargs)
    829 ids, _, _ = self.group_info
    830 ngroups = self.ngroups
--> 831 return cy_op.cython_operation(
    832     values=values,
    833     axis=axis,
    834     min_count=min_count,
    835     comp_ids=ids,
    836     ngroups=ngroups,
    837     **kwargs,
    838 )

File ~/miniconda3/envs/venv_quiche/lib/python3.9/site-packages/pandas/core/groupby/ops.py:541, in WrappedCythonOp.cython_operation(self, values, axis, min_count, comp_ids, ngroups, **kwargs)
    537 self._validate_axis(axis, values)
    539 if not isinstance(values, np.ndarray):
    540     # i.e. ExtensionArray
--> 541     return values._groupby_op(
    542         how=self.how,
    543         has_dropped_na=self.has_dropped_na,
    544         min_count=min_count,
    545         ngroups=ngroups,
    546         ids=comp_ids,
    547         **kwargs,
    548     )
    550 return self._cython_op_ndim_compat(
    551     values,
    552     min_count=min_count,
   (...)
    556     **kwargs,
    557 )

File ~/miniconda3/envs/venv_quiche/lib/python3.9/site-packages/pandas/core/arrays/categorical.py:2740, in Categorical._groupby_op(self, how, has_dropped_na, min_count, ngroups, ids, **kwargs)
   2738     if kind == "transform":
   2739         raise TypeError(f"{dtype} type does not support {how} operations")
-> 2740     raise TypeError(f"{dtype} dtype does not support aggregation '{how}'")
   2742 result_mask = None
   2743 mask = self.isna()

TypeError: category dtype does not support aggregation 'mean'
```

**Review by jranek — CHANGES_REQUESTED** [review]
Looks good Cami, just some minor bug fixes. You can feel free to merge when these changes are made, I'll take a look at the multi processing shortly

File: src/SpaceCat/features.py (hunk 1 above)
Comment (camisowers) [reply src/SpaceCat/features.py:570]: I tested it on the Stanford data you sent and didn't get the error! Could've been fixed by the segmentation label bug fix.

Comment (jranek) [reply src/SpaceCat/features.py:571]: Let's revisit after package dependency updates!

**Review by jranek — APPROVED** [review]
Looks good!

**Pull request merged**